### PR TITLE
Simplified TypeInfo, take 2: factor common code in

### PR DESCRIPTION
**Pull request by andralex**
After working for a while on https://github.com/dlang/druntime/pull/3174, I figured I was straying increasingly in breaking compatibility territory. This PR starts from the other end - the currently hardcoded typeinfo implementations are factored into common templates.

### CODE REVIEW

**Comment by MoonlightSentinel** [comment]
That's quite an impressive diff.

> After working for a while on #3174, I figured I was straying increasingly in breaking compatibility territory.

I would strongly suggest adding more tests before doing this transition because the current coverage is pretty low (https://codecov.io/gh/dlang/druntime/tree/master/src/rt/typeinfo)

**Comment by andralex** [comment]
@MoonlightSentinel yah, 2000 negative LOC is quite satisfying. I'm thinking of adding tests after this is pulled so as to not make it even more hairy. Is there a way to see coverage for this PR?

**Comment by wilzbach** [comment]
> @MoonlightSentinel yah, 2000 negative LOC is quite satisfying. I'm thinking of adding tests after this is pulled so as to not make it even more hairy. Is there a way to see coverage for this PR?

Look for CodeCov:

https://github.com/dlang/druntime/pull/3206/checks?check_run_id=1076583494
https://codecov.io/gh/dlang/druntime/pull/3206/tree

**Comment by MoonlightSentinel** [comment]
>  I'm thinking of adding tests after this is pulled so as to not make it even more hairy. 

It would be better if these tests were based on the current implementation and merged in another PR. That should catch subtle changes in behaviour far more reliably.

> Is there a way to see coverage for this PR?

Look for `codecov/patch`-> `Details` among the other CI checks.

**Comment by andralex** [comment]
Actually I saw that coverage hints were right in the diff view - really nice! Added a few, let's see how it works.

**Comment by andralex** [comment]
@MoonlightSentinel thanks for bringing coverage up... turns out that `TypeInfo_C` was dead code!

**Comment by andralex** [comment]
Bunch of changes in the generated docs because this eliminates all useless pages such as https://dlang.org/phobos/rt_typeinfo_ti_Along.html.

**Review by WalterBright — APPROVED** [review]

File: src/rt/util/typeinfo.d [hunk 1]
```diff
@@ -266,3 +266,522 @@ unittest
...
+        assert(typeid(cfloat).argTypes(t1, t2) == 0 && t1 == typeid(double) &&
+            t2 is null);
+        assert(typeid(cdouble).argTypes(t1, t2) == 0 && t1 == typeid(double) &&
+            t2 == typeid(double));
+        assert(typeid(creal).argTypes(t1, t2) == 0 && t1 == typeid(real) &&
+            t2 == typeid(real));
+    }
+
+// All pointer types are represented by `TypeInfo_P`.
+// Please keep in sync with `TypeInfo_Pointer`.
```
[the remark this answers is not in the document]
Comment (andralex) [reply src/rt/util/typeinfo.d:600]: Turns out this was dead code, too!

File: src/rt/util/typeinfo.d [hunk 2]
```diff
@@ -266,3 +266,522 @@ unittest
...
+class TypeInfo_As : TypeInfoArrayGeneric!short {}
+class TypeInfo_At : TypeInfoArrayGeneric!ushort {}
+class TypeInfo_Au : TypeInfoArrayGeneric!wchar {}
+class TypeInfo_Ai : TypeInfoArrayGeneric!int {}
+class TypeInfo_Ak : TypeInfoArrayGeneric!uint {}
+class TypeInfo_Aw : TypeInfoArrayGeneric!dchar {}
+class TypeInfo_Al : TypeInfoArrayGeneric!long {}
+class TypeInfo_Am : TypeInfoArrayGeneric!ulong {}
+
+extern (C) void[] _adSort(void[] a, TypeInfo ti);
```
[the remark this answers is not in the document]
Comment (andralex) [reply src/rt/util/typeinfo.d:645]: Hm, tried that and I get linker errors. 

```
Undefined symbols for architecture x86_64:
  "__D2rt4util8typeinfo18__unittest_L619_C1FZ7_adSortUAvC8TypeInfoZQn", referenced from:
      __D2rt4util8typeinfo18__unittest_L619_C1FZv in test_runner.o
```

[the remark this answers is not in the document]
Comment (ibuclaw) [reply src/rt/util/typeinfo.d]: `version (CoreUnittest)` then.

[the remark this answers is not in the document]
Comment (andralex) [reply src/rt/util/typeinfo.d]: you got it

[the remark this answers is not in the document]
Comment (andralex) [reply src/rt/util/typeinfo.d]: made it private, too, like a boss

**Comment by andralex** [comment]
I have a couple more thoughts, but this step is good enough. Will set auto-merge. Thanks for reviewing!

File: src/rt/util/typeinfo.d [hunk 3]
```diff
@@ -266,3 +266,520 @@ unittest
         }
     }();
 }
+
+// Reduces to `T` if `cond` is `true` or `U` otherwise.
+private template Select(bool cond, T, U)
+{
+    static if (cond) alias Select = T;
+    else alias Select = U;
+}
```
Comment (MoonlightSentinel) [inline src/rt/util/typeinfo.d:275]: Belongs into `core.internal.traits`

File: src/rt/util/typeinfo.d [hunk 4]
```diff
@@ -266,3 +266,520 @@ unittest
...
+            }
+            else
+            {
+                auto lhs = *cast(T *) p1, rhs = *cast(T *) p2;
+                return (lhs > rhs) - (lhs < rhs);
+            }
+        }
+
+    static if (is(T == Base))
+        override @property size_t tsize() nothrow pure
```
Comment (MoonlightSentinel) [inline src/rt/util/typeinfo.d:336]: Attributes are redundant

File: src/rt/util/typeinfo.d [hunk 5]
```diff
@@ -266,3 +266,520 @@ unittest
...
+    static if (is(T == Base))
+        override void swap(void *p1, void *p2)
+        {
+            auto t = *cast(T *) p1;
+            *cast(T *)p1 = *cast(T *)p2;
+            *cast(T *)p2 = t;
+        }
+
+    static if (is(T == Base) || RTInfo!T != RTInfo!Base)
+        override @property immutable(void)* rtInfo() nothrow pure const @safe
```
Comment (MoonlightSentinel) [inline src/rt/util/typeinfo.d:372]: ditto

**Review by MoonlightSentinel — COMMENTED** [review]
The module description needs to be updated as well.

On the bright side, the unittests pass with the old & new implementation.

**Comment by andralex** [comment]
@MoonlightSentinel rhanks for reviewing, see https://github.com/dlang/druntime/pull/3208

**Pull request merged**